test/integration/endpointslice/endpointslicemirroring_test.go

```diff
@@ -528,6 +528,123 @@ func TestEndpointSliceMirroringSelectorTransition(t *testing.T) {
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	// Start informer and controllers
+	stopCh := make(chan struct{})
+	closedCh := make(chan struct{})
+	informer.Start(stopCh)
+	go func() {
+		epsmController.Run(1, stopCh)
+		closedCh <- struct{}{}
+	}()
+
+	// 1. create service, endpoint, wait controller to mirror.
+	ns := framework.CreateNamespaceOrDie(client, "test-endpointslice-mirroring", t)
+	defer framework.DeleteNamespaceOrDie(client, ns, t)
+
+	service := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-123",
+			Namespace: ns.Name,
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{{
+				Port: int32(80),
+			}},
+		},
+	}
+
+	customEndpoints := &corev1.Endpoints{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        "test-123",
+			Namespace:   ns.Name,
+			Labels:      map[string]string{},
+			Annotations: map[string]string{},
+		},
+		Subsets: []corev1.EndpointSubset{{
+			Ports: []corev1.EndpointPort{{
+				Port: 80,
+			}},
+			Addresses: []corev1.EndpointAddress{{
+				IP: "10.0.0.1",
+			}},
+		}},
+	}
+
+	_, err = client.CoreV1().Services(ns.Name).Create(ctx, service, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Error creating service: %v", err)
+	}
+
+	_, err = client.CoreV1().Endpoints(ns.Name).Create(ctx, customEndpoints, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Error creating endpoints: %v", err)
+	}
+
+	// verify the expected number of mirrored slices exist
+	err = waitForMirroredSlices(t, client, ns.Name, service.Name, 1)
+	if err != nil {
+		t.Fatalf("Timed out waiting for initial mirrored slices to match expectations: %v", err)
+	}
+
+	//2. restart mirroring controller
+	close(stopCh)
+	<-closedCh
+
+	informer = informers.NewSharedInformerFactory(client, resyncPeriod)
+	epsmController = endpointslicemirroring.NewController(
+		informer.Core().V1().Endpoints(),
+		informer.Discovery().V1().EndpointSlices(),
+		informer.Core().V1().Services(),
+		int32(100),
+		client,
+		1*time.Second)
+	informer.Start(ctx.Done())
+	go epsmController.Run(1, ctx.Done())
+
+	// delete endpointslice
+	lSelector := discovery.LabelServiceName + "=" + service.Name
+	lSelector += "," + discovery.LabelManagedBy + "=endpointslicemirroring-controller.k8s.io"
+	esList, err := client.DiscoveryV1().EndpointSlices(ns.Name).List(context.TODO(), metav1.ListOptions{LabelSelector: lSelector})
+	if err != nil {
+		t.Fatalf("Error listing EndpointSlices: %v", err)
+	}
+	if len(esList.Items) == 0 {
+		t.Fatalf("expected endpoint(%s/%s) mirrored endpointslice not found", customEndpoints.Namespace, customEndpoints.Name)
+	}
+	err = client.DiscoveryV1().EndpointSlices(ns.Name).Delete(ctx, esList.Items[0].Name, metav1.DeleteOptions{})
+	if err != nil {
+		t.Fatalf("Error deleting EndpointSlices(%s/%s): %v", ns.Name, esList.Items[0].Name, err)
+	}
+	// wait endpoint to be created
+	err = waitForMirroredSlices(t, client, ns.Name, service.Name, len(esList.Items))
```

Comment:
I think that you have to replace `len(esList.Items)` by 1 here

Reply:
ah, nevermind, this has to be > 0

Reply:
this is ok

Reply:
I think first we have to assert the current slice has been deleted, otherwise we can List the slice being deleted

```diff
diff --git a/test/integration/endpointslice/endpointslicemirroring_test.go b/test/integration/endpointslice/endpointslicemirroring_test.go
index 1959872d219..993a33a0db3 100644
--- a/test/integration/endpointslice/endpointslicemirroring_test.go
+++ b/test/integration/endpointslice/endpointslicemirroring_test.go
@@ -26,7 +26,9 @@ import (
        corev1 "k8s.io/api/core/v1"
        discovery "k8s.io/api/discovery/v1"
        apiequality "k8s.io/apimachinery/pkg/api/equality"
+       apierrors "k8s.io/apimachinery/pkg/api/errors"
        metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
        "k8s.io/apimachinery/pkg/util/wait"
        "k8s.io/client-go/informers"
        clientset "k8s.io/client-go/kubernetes"
@@ -638,8 +640,22 @@ func TestEndpointSliceMirroringDeleteWhenEndpoointSliceMirroringControllerRestar
        if err != nil {
                t.Fatalf("Error deleting EndpointSlices(%s/%s): %v", ns.Name, esList.Items[0].Name, err)
        }
+       // wait endpoint slice to be deleted
+       err = wait.PollImmediate(1*time.Second, wait.ForeverTestTimeout, func() (bool, error) {
+               _, err := client.DiscoveryV1().EndpointSlices(ns.Name).Get(ctx, esList.Items[0].Name, metav1.GetOptions{})
+               if err != nil {
+                       if apierrors.IsNotFound(err) {
+                               return true, nil
+                       }
+                       return false, err
+               }
+               return false, nil
+       })
+       if err != nil {
+               t.Fatalf("Error deleting EndpointSlices(%s/%s): %v", ns.Name, esList.Items[0].Name, err)
+       }
        // wait endpoint to be created
```

Reply:
this way we avoid the race